src/CheckStrings.jl

```diff
@@ -0,0 +1,200 @@
+
+is_surrogate_lead(c::Unsigned) = ((c & ~0x003ff) == 0xd800)
+is_surrogate_trail(c::Unsigned) = ((c & ~0x003ff) == 0xdc00)
+is_surrogate_codeunit(c::Unsigned) = ((c & ~0x007ff) == 0xd800)
+is_valid_continuation(c) = ((c & 0xc0) == 0x80)
+
+## Options for check_string_* functions
+
+const UTF_NO_LONG_NULL = 1      ##< don't accept 0xc0 0x80 for '\0'
+const UTF_NO_SURROGATES = 2     ##< don't accept surrogate pairs in UTF-8/UTF-32
+const UTF_ACCEPT_LONG = 4       ##< accept long encodings (other than long null in UTF-8)
+
+const UTF_LONG = 1              ##< Long encodings are present
+const UTF_LATIN1 = 2            ##< characters in range 0x80-0xFF present
+const UTF_UNICODE2 = 4          ##< characters in range 0x100-0x7ff present
+const UTF_UNICODE3 = 8          ##< characters in range 0x800-0xd7ff, 0xe000-0xffff
+const UTF_UNICODE4 = 16         ##< non-BMP characters present
+const UTF_SURROGATE = 32        ##< surrogate pairs present
+
+## Get a UTF-8 continuation byte, give error if invalid, return updated character value
+@inline function get_continuation(ch::UInt32, byt::UInt8, pos)
+    !is_valid_continuation(byt) && throw(UnicodeError(UTF_ERR_CONT, pos, byt))
+    (ch << 6) | (byt & 0x3f)
+end
+
+CodeUnitType = Union(UInt8, UInt16, UInt32, Char)
+CodeUnitC = Union(Vector{Char}, AbstractVector{Char}, AbstractArray{Char})
+CodeUnit8 = Union(Vector{UInt8}, AbstractVector{UInt8}, AbstractArray{UInt8})
+CodeUnit16 = Union(Vector{UInt16}, AbstractVector{UInt16}, AbstractArray{UInt16})
+CodeUnit32 = Union(Vector{UInt32}, AbstractVector{UInt32}, AbstractArray{UInt32})
+CodeUnit = Union(CodeUnit8, CodeUnit16, CodeUnit32, CodeUnitC, AbstractString)
+
+"
+Validates and calculates number of characters in a UTF-8 encoded vector of `UInt8`
+
+### Input Arguments:
+* `::Type{Encoding}`
+* `dat::CodeUnit` Vector of `UInt8`, `UInt16`, `UInt32` or `Char`
+
+### Optional Input Arguments:
+* `endpos` end position (defaults to `endof(dat)`)
+* `pos`    start position (defaults to `start(dat)`)
+
+### Keyword Arguments:
+* `accept_long_null` = `true` # Modified UTF-8
+* `accept_surrogates` = `true` # CESU-8
+* `accept_long` = `false`
+
+### Returns:
+*   (total characters, flags, 4-byte, 3-byte, 2-byte)
+
+### Throws:
+*   `UnicodeError`
+"
+function check_string{T <: CodeUnit, E <: Union(UTF8, UTF16, UTF32)} (
+                      ::Type{E},
+                      dat::T,
+                      endpos = endof(dat),
+                      pos = (T <: AbstractString) ? start(dat) : 1
+                      ;
+                      accept_long_null = true,
+                      accept_surrogates = true,
+                      accept_long = false)
+    local byt::UInt8, ch::UInt32, surr::UInt32
+    flags::UInt = 0
+    totalchar = num2byte = num3byte = num4byte = 0
+    @inbounds while pos <= endpos
+        ch, pos = next(dat, pos)
+        totalchar += 1
+        if ch > 0x7f
+            if E <: UTF8
```

Comment:
Any time I find myself doing `if X <: Y`, it screams for use of dispatch. I think this is probably why @tkelman isn't super impressed by this vs. JuliaLang/julia#11575, since you've basically just added this if statement here and combined the two methods. A better solution would be to find a way to further split out the internals of `check` that could be dispatched on according to `Encoding`. This leads to nice hierarchy of methods with the most generic at the top level, and `Encoding`-specific methods at the lowest and hopefully smallest level.